src/inputs/internal/SelectFieldBase.tsx

```diff
@@ -10,25 +10,41 @@ import { ListBox } from "src/inputs/internal/ListBox";
 import { SelectFieldInput } from "src/inputs/internal/SelectFieldInput";
 import { keyToValue, Value, valueToKey } from "src/inputs/Value";
 import { BeamFocusableProps } from "src/interfaces";
+import { areArraysEqual } from "src/utils";
 
-type FieldState<O> = {
-  isOpen: boolean;
-  selectedKeys: Key[];
-  inputValue: string;
-  filteredOptions: O[];
-  selectedOptions: O[];
-};
-
-export interface SelectFieldBaseProps<O, V extends Value> extends BeamSelectFieldBaseProps<O, V> {
+export interface BeamSelectFieldBaseProps<O, V extends Value> extends BeamFocusableProps, PresentationFieldProps {
   /** Renders `opt` in the dropdown menu, defaults to the `getOptionLabel` prop. */
   getOptionMenuLabel?: (opt: O) => string | ReactNode;
   getOptionValue: (opt: O) => V;
   getOptionLabel: (opt: O) => string;
   /** The current value; it can be `undefined`, even if `V` cannot be. */
   values: V[] | undefined;
-  onSelect: (values: V[]) => void;
-  options: O[];
+  onSelect: (values: V[], opts: O[]) => void;
   multiselect?: boolean;
+  disabledOptions?: V[];
+  options: O[] | { initial: O[]; load: () => Promise<{ options: O[] }> };
```

Comment:
The Project Switcher is one case where we do something like:
**In Blueprint** - Request the project information required for displaying the current project in the Project Switcher. This will be an API request of only one option (the current project). This is what is populated in the `initial` prop. Once the user interacts with the SelectField, then the `load` event is fired. Handling the promise locally in the SelectField allows us to manage the loading state internally, vs requiring a `optionsLoading: boolean;` prop to be changed by the caller.